onvm/onvm_nflib/onvm_nflib.c

```diff
@@ -1386,3 +1500,265 @@ onvm_nflib_stats_summary_output(uint16_t id) {
+                        }
+                        count_sleep++;
+                        sleep(1);
+                }
+                if (rte_ring_enqueue(nf_pool_ring, (void *) spawned_nf) != 0) {
+                        RTE_LOG(INFO, APP, "Failed to enqueue spawned NF into the ring\n");
+                        break;
+                }
+                spawned_nf->pool_status.pool_sleep_state = 1;
+        }
+
+        return i;
+}
+
+int
+onvm_nflib_pool_dequeue(const char *nf_name, int dq_num, int refill) {
+        struct rte_ring *nf_pool_ring;
+        struct onvm_nf_pool_ctx *pool_ctx;
+        struct onvm_nf *dequeued_nf = NULL;
+        sem_t *sem;
+        int i, ret, total_refill, spawned_nf_count;
+        char *global_nf_name;
+
+        global_nf_name = rte_calloc(NULL, 1, 64, 0);
+        snprintf(global_nf_name, 64, "%s", nf_name);
+        ret = rte_hash_lookup_data(pool_map, global_nf_name, (void *) &pool_ctx);
+        if (ret < 0) {
+                RTE_LOG(INFO, APP, "NF does not have an associated pool\n");
+                return -1;
+        }
+
+        nf_pool_ring = pool_ctx->pool_ring;
+        for (i = 0; i < dq_num; i++) {
+                if (rte_ring_dequeue(nf_pool_ring, (void *) &dequeued_nf) != 0) {
+                        RTE_LOG(INFO, APP, "Could not dequeue %s, ring is empty\n", nf_name);
+                        break;
+                }
+                sem = sem_open(dequeued_nf->pool_status.pool_mutex_name, 0, 0666, 0);
+                dequeued_nf->pool_status.pool_sleep_state = 0;
+                if (sem_post(sem) < 0) {
+                        RTE_LOG(INFO, APP, "Could not post to semaphore\n");
+                        break;
+                }
+        }
+        if (refill >= 0) {
+                pool_ctx->refill = refill;
+        }
+
+        if (rte_ring_count(nf_pool_ring) < pool_ctx->refill) {
+                total_refill = pool_ctx->refill - rte_ring_count(nf_pool_ring);
+                spawned_nf_count = onvm_nflib_fork_pool_nfs(pool_ctx->nf_name, pool_ctx->args,
+                                                            nf_pool_ring, total_refill);
+                RTE_LOG(INFO, APP, "Refilled NF pool with %d NF's\n", spawned_nf_count);
+        }
+
+        return i;
+}
+
+struct onvm_nf *
+onvm_nflib_fork(const char *nf_name, void *nf_args) {
+        struct simple_forward_args *args_sf;
+        struct aes_decrypt_args *args_aesdec;
+        char *binary_string;
+        int fork_error;
+        int instance_id;
+
+        if (nf_name == NULL) {
+                RTE_LOG(INFO, APP, "onvm_nflib_fork(): nf_name is null\n");
+                return NULL;
+        }
+        if (nf_args == NULL) {
+                RTE_LOG(INFO, APP, "onvm_nflib_fork(): nf_args is null\n");
+                return NULL;
+        }
+
+        instance_id = onvm_nflib_request_next_instance_id();
+        binary_string = onvm_nflib_create_binary_exec_string(nf_name);
+        if (fork() == 0) {
```

Comment:
Yeahhhh so this whole hard coded arg thing just seems hacky even without reading too much into it. Can you explain what we're trying to achieve here

Reply:
It is hacky and should be more automated and clean. The idea is that each NF gets its executable path,and then we do an execl to load that process, starting the NF. Getting the args to the Nf's isn't straightforward - in this case a struct is used to hold the arguments to the Nf. A user can specify many combinations of the arguments and thats why this branch is so long. This could be replaced with something like calling a shell script from the C code to let it do the messy business of starting the NF.

Reply:
Okay stick something info the nf infromation which identifies the execl format string and define the rqeuired args for it. That was we can be NF agnostic in this

Reply:
I refactored to directly call the NF bash script.